### ISSUE TITLE
TestAPI AddTestResultsToTestRun always failing due to undocumented fields.

### ISSUE THREAD

**Issue by Coderrob** [microsoft/azure-devops-node-api#318]
### Environment
Node version: 8.12.0 
Npm version: 6.4.1
OS and version: Win10
azure-devops-node-api version: ^8.0.0

### Issue Description
Calling `testApi.addTestResultsToTestRun([testResult], projectName, testRunId)` bombs regardless of what shallow copies or values are provided. I've tried the payload for the result:

```
const testResult = {
        testPlan: { id: planId },
        testSuite: { id: suiteId },
        testCase: { id: testCaseId },
        testPoint: { id: pointId },
        state: 'Completed',
        outcome, // ex: 'Passed'
      };
```
Keep running into this error:

![errors](https://user-images.githubusercontent.com/7213776/58903417-014fc980-86cb-11e9-9553-39a00582aabc.PNG)

Following the source code (actual documentation would be _super_) I can see the following function fields that are required to make the call. 

https://github.com/microsoft/azure-devops-node-api/blob/9ecc84ccfca7968b6ed8d5499c39d59a9e5ad8d2/api/TestApi.ts#L60

`addTestResultsToTestRun(results: TestInterfaces.TestCaseResult[], project: string, runId: number): Promise<TestInterfaces.TestCaseResult[]>;`

If I follow the interface for `TestCaseResult` here:

https://github.com/microsoft/azure-devops-node-api/blob/9ecc84ccfca7968b6ed8d5499c39d59a9e5ad8d2/api/interfaces/TestInterfaces.ts#L2340

There is no mention of `TestsPointId` or other fields that are required. So, i'm not convinced that this function actually works. I've followed along with the proxy open for Fiddler to see the payloads being sent and everything looks to match the required fields, but there is still the error.

### Expected behaviour
I should be able to add the test results to a test run successfully. Right now I'm getting required field assertions for fields not defined in the interface for the function's contract. 

I've setup test reporting around the 5.0x REST Api, but trying to use this repo's code has been a failure. The provided sample for this project only goes as far as creating a test run then nothing else. 

### Actual behaviour
The call to `addTestResultsToTestRun` fails asserting that `TestPointId, testCaseId, testCaseRevision, testCaseTitle must be specified for planned test results.`... notice the casing is also inconsistent in the message as well.

### Steps to reproduce
1. Create a TestApi service object using a known/valid personal access token
2. Create a new test run `testApi.createTestRun`
3. Call `testApi.getPoints` to get the point id for a test case
4. Call `testApi.addTestResultsToTestRun` to add passed/failed results.
5. ???
6. Profit

### Logs
```
Error: TestPointId, testCaseId, testCaseRevision, testCaseTitle must be specified for planned test results.
message:"TestPointId, testCaseId, testCaseRevision, testCaseTitle must be specified for planned test results."
result:Object {$id: "1", innerException: null, message: "TestPointId, testCaseId, testCaseRevision, testCas…", …}
$id:"1"
errorCode:0
eventId:3000
innerException:null
message:"TestPointId, testCaseId, testCaseRevision, testCaseTitle must be specified for planned test results."
typeKey:"InvalidPropertyException"
typeName:"Microsoft.TeamFoundation.TestManagement.WebApi.InvalidPropertyException, Microsoft.TeamFoundation.TestManagement.WebApi"
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
stack:"Error: TestPointId, testCaseId, testCaseRevision, testCaseTitle must be specified for planned test results.\n    at RestClient.<anonymous> (c:\Devspace\bounty-hunters\node_modules\typed-rest-client\RestClient.js:200:31)\n    at Generator.next (<anonymous>)\n    at fulfilled (c:\Devspace\bounty-hunters\node_modules\typed-rest-client\RestClient.js:6:58)\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:189:7)"
statusCode:400
__proto__:Object {constructor: , name: "Error", message: "", …}
constructor:function Error() { … }
message:""
name:"Error"
toString:function toString() { … }
__proto__:Object {constructor: , __defineGetter__: , __defineSetter__: , …}
```
```

**Comment by Coderrob** [comment]
I found a workaround until this issue is resolved by overriding the function `addTestResultsToTestRun` and downdgrading the api version to `'5.0-preview.5'`. 

```
const addReportingOverride = (api) => {
  // eslint-disable-next-line func-names
  api.addTestResultsToTestRun = function (results, projectName, runId) {
    return new Promise(async (resolve, reject) => {
      const routeValues = {
        project: projectName,
        runId,
      };

      try {
        const verData = await this.vsoClient.getVersioningData(
          '5.0-preview.5',
          'Test',
          '4637d869-3a76-4468-8057-0bb02aa385cf',
          routeValues,
        );

        const url = verData.requestUrl;
        const options = this.createRequestOptions('application/json', verData.apiVersion);
        const res = await this.rest.create(url, results, options);
        resolve(res);
      } catch (err) {
        reject(err);
      }
    });
  };
  return api;
};
```

**Comment by damccorm** [comment]
Cool, that's helpful. I'm guessing that it probably just needs to be regenerated then, it must have gotten out of sync with the endpoint somehow. I'm planning on regening the whole library soon (tonight or tomorrow morning), so hopefully that should fix the issue.

**Comment by damccorm** [comment]
As far as the larger point about docs goes, these APIs have pretty extensive docs [here](https://docs.microsoft.com/en-us/rest/api/vsts/?view=vsts-rest-4.1), including one for [this function](https://docs.microsoft.com/en-us/rest/api/azure/devops/test/results/add?view=azure-devops-rest-5.1).

This library is just generated from those endpoints, so while we can pull in comments in the source code there's not a great way for us to provide more formal docs within the context of this library.

**Comment by Coderrob** [comment]
@damccorm I've read through the documentation a few times from 4/5.0/5.1. A lot of MS documentation pages/sites there is often github integration with repo monitors, but not the case for ADO REST API docs.

There are no indicators (I've seen anyway) to know what is a required field or combo of fields with the functions.

I was hoping to see a sample here of expected calls using the TestAPI from creating a test run, submitting results, and then ending the test run. Still finding that you have to submit the `endTestRun` twice to get ADO to see that the test run is complete.

I'm hoping to just consume this code so I don't have to keep maintaining my own implementations moving forward. Hoping the next re-gen can correct that, but I'm thinking it may be a failing in the REST API itself and not this project.

**Comment by damccorm** [comment]
> A lot of MS documentation pages/sites there is often github integration

I think this actually does exist, see the feedback section at the bottom of the page. That in turn links to [this repo](https://github.com/MicrosoftDocs/vsts-rest-api-specs/issues?utf8=%E2%9C%93&q=%22f74b7d51-aede-ed93-437e-601654d8fa7e%22&in=body)

> Hoping the next re-gen can correct that, but I'm thinking it may be a failing in the REST API itself

Just added a PR to regen (#320), it looks like its probably not going to fix this unfortunately, so I'm guessing you're right that its an issue with the endpoint itself. Probably the best path forward is to file an issue either in the repo I linked to earlier or add a ticket on the [developer community](https://developercommunity.visualstudio.com/spaces/21/index.html) site - they'll know the appropriate place to route it. Sorry we can't be more help in this library 😕

**Comment by Coderrob** [comment]
This library got me about 90% of the way towards not having to use my own implementation, and I was able to override the results function to make it work. That's still a win for me. Thanks @damccorm!

**Comment by damccorm** [comment]
Glad its helpful, hopefully these endpoint issues get sorted out as well.

I'm going to close here because I dont _think_ we have anything left to track on our end. If I'm mistaken just comment and I'll reopen.

**Comment by Azka-Gilani** [comment]
@satyen95 This can be resolved by using 16.143.1 version of Microsoft.TeamFoundationsServer.ExtendedClient. This error also doesn't appear in 15.131.1 but that has framework restrictions.
